Automattic/jetpack
# Blocks: Add Facebook and Instagram Embed block variations

"Successor" to D49605-code and 522-gh-wpcomsh. Once this is merged to JP, and the corresponding WP.com diff (D49756-code) is committed to WP.com, we can retire those changes.

#### Changes proposed in this Pull Request:

Quoting 522-gh-wpcomsh:

> Register a callback for the `block.registerBlockType` frontend filter that patches the `Facebook` and `Instagram` `core/embed` block variations so they appear in the inserter and can have their relevant URLs parsed again.
>
> This is relevant for **Gutenberg >= 9.0** (see [CfT](p7DVsv-9jU-p2#comment-31039)), where the aforementioned blocks [were deprecated](https://github.com/WordPress/gutenberg/pull/24472) because they do not support the oEmbed changes that Facebook is going to implement on Oct. 24th, 2020, and core has currently no plans to support it.
>
>  However, we do plan on keeping support for them on WPCOM/AT. Facebook is adding a token authentication mechanism to its new oEmbed implementation, and we are following suit.
>
>Our goal is for this go unnoticed by our end-users, as this is only an implementation detail.

#### Jetpack product discussion
p7DVsv-9jU-p2#comment-31116

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
Make sure you're testing on a WP install with Gutenberg 9.0 installed.

This diff is rather simple -- the crucial part is to make sure that we're actually not making any more requests to the legacy API endpoints for Instagram oEmbeds (https://api.instagram.com/oembed). We'll be testing that quite thoroughly in the following.

The Facebook embeds OTOH are largely unaffected, since they haven't been previewed in the editor before (instead, we've been displaying an `Embedded content from facebook.com can't be previewed in the editor.` message in the block). (Some discussion on whether that's the way it _should_ be here: https://github.com/Automattic/wp-calypso/issues/27701#issuecomment-693351019, but that's not really relevant to this PR.)

- We need to direct the REST API requests your local Jetpack dev install makes to your WP.com sandbox: In your local Jetpack dev env, sandbox the REST API, e.g. by adding a `0-sandbox.php` file to `docker/mu-plugins/`, with the following contents:

```php
<?php

define( 'JETPACK__SANDBOX_DOMAIN', 'YOURSANDBOX' );
```

(change `YOURSANDBOX` to your sandbox' URL). Make sure to restart your JP install afterwards. For more information on this, see PCYsg-efV-p2 (Section "The `JETPACK__SANDBOX_DOMAIN` PHP Constant").

- Make sure that your local JP install is really using your sandbox for these requests: Log out of your sandbox, and paste the link of an Instagram post into your JP install's Gutenberg editor. You should get an empty embed block showing a notice saying something like "Sorry, this content could not be embedded".
- Log into your sandbox, and verify that pasting an IG post URL is now successfully transformed into an embed. _Use a different IG post to avoid getting a cached result._ 
- Apply the following patch to your sandbox to [simulate](https://developers.facebook.com/docs/instagram/oembed-legacy) the retiring of Instagram's legacy API endpoint:

```diff
Index: wp-content/rest-api-plugins/endpoints/oembed-proxy.php
===================================================================
--- wp-content/rest-api-plugins/endpoints/oembed-proxy.php	(revision 213803)
+++ wp-content/rest-api-plugins/endpoints/oembed-proxy.php	(working copy)
@@ -52,8 +52,10 @@
 				'https://graph.facebook.com/v5.0/instagram_oembed'
 			);
 		} else {
+			$query_args = $request->get_query_params();
+			$query_args['breaking_change'] = 'oembed';
 			$request_url = add_query_arg(
-				$request->get_query_params(),
+				$query_args,
 				'https://api.instagram.com/oembed/'
 			);
 		}
```
- Try inserting another IG post URL (again a different one, because caching). It should fail.
- Fortunately, we already have code in place to use the new oEmbed API instead. To enable it, set

```
$auth_to_instagram = true;
```

on L40 of `wp-content/rest-api-plugins/endpoints/oembed-proxy.php` in your sandbox.
(I can't post a diff here, since it would include an internal link.)

- Verify that the block embed works now.
- Finally, we need to simulate WP.com behavior after the legacy oEmbed API has been retired. To that end, apply the following patch _to your local JP install_ (and reload the block editor):

```diff
diff --git a/modules/shortcodes/instagram.php b/modules/shortcodes/instagram.php
index 8a24c479b..bb80ed709 100644
--- a/modules/shortcodes/instagram.php
+++ b/modules/shortcodes/instagram.php
@@ -248,7 +248,8 @@ function jetpack_instagram_use_cache( $matches, $atts, $passed_url ) {
  * @return mixed An object if successful or a WP_Error object
  */
 function jetpack_instagram_fetch_embed( $args ) {
-       if ( Constants::is_defined( 'IS_WPCOM' ) && Constants::get_constant( 'IS_WPCOM' ) ) {
+       if ( true ) {
+               $args['breaking_change'] = 'oembed';
                $url      = esc_url_raw(
                        add_query_arg(
                                $args,

```
- Verify that pasting an IG URL doesn't work now. This means that we'll have to modify the `if` clause to make sure we're also using the new API for WP.com after Oct 24.
- Remove all patches that you've applied on top of this branch, or on your sandbox. Disconnect your JP from WP.com (at `/wp-admin/admin.php?page=jetpack#/dashboard`, "Manage website connection"). Turn on [Offline Mode](https://jetpack.com/support/development-mode/). Reload the editor, and verify that the IG embed is gone from the inserter, while the FB one is still there.

#### Conclusion from Testing Instructions

- The blocks added by this PR currently work -- the PR can be merged as-is :heavy_check_mark: 
- To prepare for Facebook's retiring of the old oEmbed API, we need to set `$auth_to_instagram = true` in `wp-content/rest-api-plugins/endpoints/oembed-proxy.php` before Oct 24. (We know this much is `true`.)
- To also make sure that the block variations work on WP.com after Oct 24, we need to file a PR that modifies the conditional in `jetpack_instagram_fetch_embed` in [`modules/shortcodes/instagram.php`](https://github.com/Automattic/jetpack/blob/8f8104179df236101209af8f71aef3e70d06d340/modules/shortcodes/instagram.php) accordingly.

#### Proposed changelog entry for your changes:
Add Facebook and Instagram Embed block variations.

cc/ @Automattic/good-mountain @ebinnion @jeherve

Fixes https://github.com/Automattic/wp-calypso/issues/45832
